### ISSUE TITLE
K8s 1.11.6 fails to install : Requires: kubernetes-cni = 0.6.0

### ISSUE THREAD

**Issue by bfurey** [kubernetes/kubernetes#75701]
**What happened**:

When deploying K8s 1.11.6, we get an error: Requires: kubernetes-cni = 0.6.0
 - we are required to lock down on K8s 1.11.6 and thus cannot move to K8s 1.11.9

```
/usr/bin/yum --noplugins install -y kubelet-1.11.6* kubeadm-1.11.6* kubectl-1.11.6* --disableexcludes=kubernetes

Resolving Dependencies
--> Running transaction check
---> Package kubeadm.x86_64 0:1.11.6-0 will be installed
--> Processing Dependency: kubernetes-cni >= 0.6.0 for package: kubeadm-1.11.6-0.x86_64

---> Package kubectl.x86_64 0:1.11.6-0 will be installed
---> Package kubelet.x86_64 0:1.11.6-0 will be installed
--> Processing Dependency: kubernetes-cni = 0.6.0 for package: kubelet-1.11.6-0.x86_64
--> Running transaction check
---> Package kubelet.x86_64 0:1.11.6-0 will be installed
--> Processing Dependency: kubernetes-cni = 0.6.0 for package: kubelet-1.11.6-0.x86_64
---> Package kubernetes-cni.x86_64 0:0.7.5-0 will be installed

--> Finished Dependency Resolution
Error: Package: kubelet-1.11.6-0.x86_64 (kubernetes)
           Requires: kubernetes-cni = 0.6.0
           Available: kubernetes-cni-0.3.0.1-0.07a8a2.x86_64 (kubernetes)
               kubernetes-cni = 0.3.0.1-0.07a8a2
           Available: kubernetes-cni-0.5.1-0.x86_64 (kubernetes)
               kubernetes-cni = 0.5.1-0
           Available: kubernetes-cni-0.5.1-1.x86_64 (kubernetes)
               kubernetes-cni = 0.5.1-1
           Available: kubernetes-cni-0.6.0-0.x86_64 (kubernetes)
               kubernetes-cni = 0.6.0-0
           Installing: kubernetes-cni-0.7.5-0.x86_64 (kubernetes)
               kubernetes-cni = 0.7.5-0
 You could try using --skip-broken to work around the problem
```

**What you expected to happen**:

K8s 1.11.6 to install clean

**How to reproduce it (as minimally and precisely as possible)**:
/usr/bin/yum --noplugins install -y kubelet-1.11.6* kubeadm-1.11.6* kubectl-1.11.6* --disableexcludes=kubernetes

**Anything else we need to know?**:
 - we are required to lock down on K8s 1.11.6 and thus cannot move to K8s 1.11.9

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.6
- Cloud provider or hardware configuration: on-premise
- OS (e.g: `cat /etc/os-release`): NAME="Red Hat Enterprise Linux Server" VERSION="7.6 (Maipo)"
- Kernel (e.g. `uname -a`): x86_64 GNU/Linux
- Install tools:
- Others:

**Comment by pswica** [comment]
I do recall that this is a dependency for kubelet. I've seen it outlined a few places:

https://github.com/kubernetes/kubernetes/issues/57855
https://github.com/kubernetes/release/issues/97

**Comment by bfurey** [comment]
Installation of K8s 1.11.6 worked up until Mar 24th 2019, which was release day for 1.14, 1.13.5, 1.12.7 and 1.11.9), so this issue would be a regression.

**Comment by neolit123** [comment]
@bfurey in all kubelet packages all the way back to 1.11 the CNI was bumped to 0.7.5 because it includes a fix for a security exploit. the bug is of medium to high severity.

so in any case you want to use CNI > 0.7.x.

here is the patch that changed that:
https://github.com/kubernetes/release/commit/55e67316071f8dfc5c8c89962a1321630c6430c7

i tried reproducing this problem on Fedora 30 and it doesn't fail for me:
<details>
Installed:
  kubeadm-1.11.6-0.x86_64                                                       
  kubectl-1.11.6-0.x86_64                                                       
  kubelet-1.11.6-0.x86_64                                                       
  iproute-tc-5.0.0-2.fc30.x86_64                                                
  iproute-5.0.0-2.fc30.x86_64                                                   
  containernetworking-plugins-0.7.4-2.fc30.x86_64                               
  ebtables-2.0.10-31.fc30.x86_64                                                
  ethtool-2:4.17-3.fc30.x86_64                                                  
  iptables-1.8.0-5.fc30.x86_64                                                  
  libmnl-1.0.4-9.fc30.x86_64                                                    
  libnetfilter_conntrack-1.0.7-2.fc30.x86_64                                    
  libnfnetlink-1.0.1-15.fc30.x86_64                                             
  linux-atm-libs-2.5.1-21.fc29.x86_64                                           
  socat-1.7.3.2-9.fc30.x86_64                                                   
  cri-tools-1.12.0-0.x86_64                                                     

Complete!
</details>

is there a workaround for your problem.

**Comment by bfurey** [comment]
@neolit123 - My current workaround is to explicitly install kubernetes-cni-0.6
>  install -y kubelet-1.11.6* kubeadm-1.11.6* kubectl-1.11.6* kubernetes-cni-0.6.*

But if 1.11.6 now should be on cni > 0.7.x, why is cni > 0.7.x not deployed as a dependency?

**Comment by neolit123** [comment]
> But if 1.11.6 now should be on cni > 0.7.x, why is cni > 0.7.x not deployed as a dependency?

i do no know. this is the PR that made the change and it was approved by multiple people:
https://github.com/kubernetes/release/pull/683

**Comment by neolit123** [comment]
> My current workaround is to explicitly install kubernetes-cni-0.6
/usr/bin/yum install -y kubelet-1.11.6* kubeadm-1.11.6* kubectl-1.11.6* kubernetes-cni-0.6.*

seems like a valid workaround to me.

**Comment by bfurey** [comment]
@neolit123 - How would an existing customer on 1.11.6 (with cni-0.6-0) perform a zero downtime upgrade to 1.11.6 (with cni-0.7.5)?

**Comment by neolit123** [comment]
the kubernetes-cni packages installs some binaries in `/opt/cni/bin`. these are the so called CNI plugins.

so if you install a Pod network solution like WeaveNet, Flannel it may or may not depend on these CNI plugins from kubernetes-cni.

i believe most of the popular ones do depend.

so you cannot upgrade CNI without the absolute zero downtime. to my knowledge, you need to stop the kubelet, remove kubernetes-cni 0.6.0, install kubernetes-cni 0.7.5, restart the kubelet.

but if you have a HA setup with multiple control plane nodes and multiple worker nodes, you can avoid the downtime if you apply the fix one node at a time.

**Comment by gertvdijk** [comment]
Also the APT repository is affected. It seems someone changed the version string of `kubernetes-cni` from `0.6.0` to `0.6.0-00` without actually updating the dependency fields in packages like `kubeadm`.

Breakdown:

```
The following packages have unmet dependencies:
 kubeadm : Depends: kubernetes-cni (= 0.6.0) but 0.7.5-00 is to be installed
 kubelet : Depends: kubernetes-cni (= 0.6.0) but 0.7.5-00 is to be installed
E: Unable to correct problems, you have held broken packages.
```

```
$ apt policy kubernetes-cni
kubernetes-cni:
  Installed: 0.7.5-00
  Candidate: 0.7.5-00
  Version table:
 *** 0.7.5-00 500
        500 http://apt.kubernetes.io kubernetes-xenial/main amd64 Packages
        100 /var/lib/dpkg/status
     0.6.0-00 500
        500 http://apt.kubernetes.io kubernetes-xenial/main amd64 Packages
[...]
```

```
# apt show kubeadm=1.12.6-00 | grep ^Depends:
Depends: kubelet (>= 1.6.0), kubectl (>= 1.6.0), kubernetes-cni (= 0.6.0), cri-tools (>= 1.12.0)
```

And `= 0.6.0` does not match to any version currently available.

Please fix ASAP.

**Comment by neolit123** [comment]
it's not that simple. the k8s package management has a problem where package dependencies are unbranched and non-atomic. so if the CNI is bumped to 0.7.5 for 1.11.9, 1.11.6 should not be affected, but it is...

there are developers assigned to work on this problem but the timeline is unclear.

**Comment by gertvdijk** [comment]
@neolit123 

Thanks for ACKing this and the update on the matter.

> it's not that simple. the k8s package management has a problem where package dependencies are unbranched and non-atomic. so **if the CNI is bumped** to 0.7.5 for 1.11.9, 1.11.6 should not be affected, but it is...

Please note that the CNI dependency was *not* bumped for older versions, yet the version strings don't match (`0.6.0` !- `0.6.0-00`). Let me rephrase my earlier point then: never delete packages in the repository once published.

Generally, if I update an APT repository, I would only _add_ packages of other versions. Deleting them (e.g. `kubernetes-cni` `0.6.0` is gone) is pretty much _not done_. If you want to refactor the version strings, that's fine, but either:
* keep the package with the older version string in, effectively publishing both `0.6.0` and `0.6.0-00`.
* create a metapackage for the dependencies or something that provides the compatibility indirection.
* rebuild *all* packages with new version strings for the dependencies (but I hope that a route you don't want to take)

**Comment by neolit123** [comment]
>  Let me rephrase my earlier point then: never delete packages in the repository once published.

i don't think that ever happened.

read this comment:
https://github.com/kubernetes/kubernetes/issues/75683#issuecomment-476750981

>  Generally, if I update an APT repository, I would only add packages of other versions. 

if we have to update the old packages, we technically can't.
we can however push revisions - e.g. `1.13.4-1`. where `-1` is the added revision.

**Comment by gertvdijk** [comment]
@neolit123 

> i don't think that ever happened.

Oh, wow, you're right. I've gone through my APT history of installs and I've noticed something worse.

Before yesterday, `1.12.x-00` depended on `kubernetes-cni:amd64 (0.6.0-00, automatic)`, but the very same package `1.12.x-00` is _republished_ with the wrong dependency (`0.6.0`).

Every APT tool I know will prevent you from publishing a package with the same name+version with different contents/hash.

This seems like broken at multiple levels at least.

**Comment by neolit123** [comment]
> Before yesterday, 1.12.x-00 depended on kubernetes-cni:amd64 (0.6.0-00, automatic), but the very same package 1.12.x-00 is republished with the wrong dependency (0.6.0).

if the packages are republished we have a problem.
this should not be ever done.

/assign @calebamiles @tpepper @justaugustus

**Comment by rmb938** [comment]
I am having the same problem with 1.13.4

```
# yum install --enablerepo=kubernetes kubelet-1.13.4-0 kubeadm-1.13.4-0 kubectl-1.13.4-0 cri-tools
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
 * elrepo: mirrors.colocall.net
 * elrepo-kernel: mirrors.colocall.net
Resolving Dependencies
--> Running transaction check
---> Package cri-tools.x86_64 0:1.12.0-0 will be installed
---> Package kubeadm.x86_64 0:1.13.4-0 will be installed
--> Processing Dependency: kubernetes-cni >= 0.6.0 for package: kubeadm-1.13.4-0.x86_64
---> Package kubectl.x86_64 0:1.13.4-0 will be installed
---> Package kubelet.x86_64 0:1.13.4-0 will be installed
--> Processing Dependency: kubernetes-cni = 0.6.0 for package: kubelet-1.13.4-0.x86_64
--> Running transaction check
---> Package kubelet.x86_64 0:1.13.4-0 will be installed
--> Processing Dependency: kubernetes-cni = 0.6.0 for package: kubelet-1.13.4-0.x86_64
---> Package kubernetes-cni.x86_64 0:0.7.5-0 will be installed
--> Finished Dependency Resolution
Error: Package: kubelet-1.13.4-0.x86_64 (kubernetes)
           Requires: kubernetes-cni = 0.6.0
           Available: kubernetes-cni-0.3.0.1-0.07a8a2.x86_64 (kubernetes)
               kubernetes-cni = 0.3.0.1-0.07a8a2
           Available: kubernetes-cni-0.5.1-0.x86_64 (kubernetes)
               kubernetes-cni = 0.5.1-0
           Available: kubernetes-cni-0.5.1-1.x86_64 (kubernetes)
               kubernetes-cni = 0.5.1-1
           Available: kubernetes-cni-0.6.0-0.x86_64 (kubernetes)
               kubernetes-cni = 0.6.0-0
           Installing: kubernetes-cni-0.7.5-0.x86_64 (kubernetes)
               kubernetes-cni = 0.7.5-0
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```

**Comment by ygalblum-stratoscale** [comment]
We're experiencing the same issue on CentOS for 1.11, 1.12, 1.13 since about the day that 1.14 was released. Our workaround was to pin the CNI version to 0.6.0 (as suggested above). But, reading this thread, I couldn't understand if it is recommended to use 0.6.0 or 0.7.5.

If it's 0.6.0, it seems that pinning the version is the way to go. But, it it's to use 0.7.5 since the installation fails, an actual solution is required

**Comment by Angha** [comment]
We're experiencing the same issue on CentOS for 1.12, 1.13   and  currently  blocked  because of this. When  will the  fix/workaround  will be available?

**Comment by neolit123** [comment]
the PR with the fix is currently blocked by reviewers.

1.12.7, 1.13.5 and 1.14.0 should not have the problem.
if you want to use *different* k8s versions, try pining the CNI version by passing `kubernetes-cni-0.6.0-0`
and then installing the `kubernetes-cni-0.7.0-0` package on top of it.

**Comment by jimmycuadra** [comment]
Any updates on this? Trying to update from k8s 1.12 to 1.13 with kubeadm and the very first step [according to the documentation](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade-1-13/) completely fails:

```
$ sudo apt-get install -y kubeadm=1.13.5-00
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 kubelet : Depends: kubernetes-cni (= 0.6.0) but 0.7.5-00 is to be installed
E: Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
```

Shouldn't this be critically urgent?

**Comment by neolit123** [comment]
try:
```
sudo apt-get install -y kubeadm=1.13.5-00 kubernetes-cni=0.6.0-00
```

and then:
```
sudo apt-get install -y kubernetes-cni=0.7.5-00
```

> Shouldn't this be critically urgent?

progress on this issue is blocked by deeper problems. packages are claimed as "best effort" until further notice.

**Comment by rmb938** [comment]
@neolit123 Is there anything we can do to make progress on this faster and packages not be "best effort"? Many companies rely on these packages working for production clusters and having to do hacks like these to fix this isn't the best.

It is very concerning that this is not top priority and what ever is blocking this isn't fixed. This kinda leaves a bad taste that OS packages aren't really supported by Kubernetes.

**Comment by neolit123** [comment]
the package specs are in this repo:
https://github.com/kubernetes/release

patch attempts for an immediate fix were rejected due to the fact that new revisions need to be pushed of old packages, which is undesired by the release maintainers.

there is a proposal to move the packages from k/release to k/kubernetes:
https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/kubeadm/20190227-artifact-generation.md
the work there is on hold for unknown to me reasons.

you can try starting a discussion at this google group:
https://groups.google.com/forum/#!forum/kubernetes-sig-release

**Comment by jimmycuadra** [comment]
No dice:

```
$ sudo apt-get install -y kubeadm=1.13.5-00 kubernetes-cni=0.6.0-00
Reading package lists... Done
Building dependency tree
Reading state information... Done
kubernetes-cni is already the newest version (0.6.0-00).
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 kubeadm : Depends: kubernetes-cni (= 0.7.5) but 0.6.0-00 is to be installed
E: Unable to correct problems, you have held broken packages.
```

**Comment by gertvdijk** [comment]
@neolit123 

> patch attempts for an immediate fix were rejected due to the fact that **new revisions need to be pushed of old packages**, which is undesired by the release maintainers.

Say what? That's how it got broken in the first place as shown in my [earlier comment](https://github.com/kubernetes/kubernetes/issues/75701#issuecomment-476757653)!

**Comment by neolit123** [comment]
@gertipoppel 

> Say what? That's how it got broken in the first place as shown in my earlier comment!

feel free to comment in here:
https://groups.google.com/forum/#!topic/kubernetes-sig-release/zkYkzzMGjic

**Comment by rmb938** [comment]
I posted a comment on that topic (pending moderator approval apparently) however it has been open since April 3rd and hasn't hasn't had any discussion since the 8th.  It is now April 16th with this github issue being open since March 25th. 

This issue **MUST** be moved forward, if it is not we are going to start seeing 3rd party packages being built.

**Comment by Anmar-Salih** [comment]
I don't see update, I am wondering if this issue has been fixed ?

**Comment by cnighojkar** [comment]
We faced same issue with v1.12.4 , when the installation for v1.12.4 will be unblocked.

**Comment by Anmar-Salih** [comment]
The problem has been fixed two days ago. Close it please.

**Comment by andrewsav-datacom** [comment]
It's still a problem on ubuntu 16.

**Comment by Anmar-Salih** [comment]
> It's still a problem on ubuntu 16.04 as described [here](https://github.com/kubernetes/kubernetes/issues/75701#issuecomment-476625403)

I used it too today on Ubuntu 16.04 and it was working smoothly. ( you have to use it on multicore processor machine in order to work) I am using it on virtualbox. So I set the VM processor number to 2.

**Comment by andrewsav-datacom** [comment]
Yeah, I tried today too and no dice. Not sure what multicore has to do with this but I also tried with 2 sockets.

**Comment by Anmar-Salih** [comment]
> It's still a problem on ubuntu 16.04 as described [here](https://github.com/kubernetes/kubernetes/issues/75701#issuecomment-476625403)

attach the error log file please

**Comment by P1ng-W1n** [comment]
> The problem has been fixed two days ago. Close it please.

Could you please point to the exact commit that's supposed to fix this issue please?

**Comment by Anmar-Salih** [comment]
> > The problem has been fixed two days ago. Close it please.
> 
> Could you please point to the exact commit that's supposed to fix this issue please?
check out this link please.

[https://review.opendev.org/#/c/659672/](url)

**Comment by P1ng-W1n** [comment]
> > > The problem has been fixed two days ago. Close it please.
> > 
> > 
> > Could you please point to the exact commit that's supposed to fix this issue please?
> > check out this link please.
> 
> https://review.opendev.org/#/c/659672/

I'm not sure that we're talking about the same issue here.. 
This thread is about an installation failure caused by packaging issues (deb, rpm OS packages).. 
The commit you highlighted is related to Devstack (development version of Openstack) and one of it's Ansible playbook, where they bumped up the k8s version to 1.13.6 (with calico 3.3).  - In short, the one has nothing to do with the other IMHO.

**Comment by Anmar-Salih** [comment]
I thought only Devstack effected by this issue. Sorry, my bad.

**Comment by andrewsav-datacom** [comment]
Yep, it looks to me too that Anmar-Salih is talking about something else, not the issue at hand. For what it worth, here is output on ubuntu:

```bash
$ sudo apt-get install -y kubelet=1.11.6-00 --allow-downgrades
Reading package lists... Done
Building dependency tree
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 kubelet : Depends: kubernetes-cni (= 0.6.0) but 0.7.5-00 is to be installed
E: Unable to correct problems, you have held broken packages.
```

Similar, to what is already reported [here](https://github.com/kubernetes/kubernetes/issues/75701#issuecomment-476625403).

**Comment by justaugustus** [comment]
Dependency comparison operators have long since been fixed for our deb/rpm packages.
For anyone still experiencing issues, please ensure you're using the most recent patch release of the Kubernetes releases that are in support.

As of writing, the following patch releases are most recent:
- https://github.com/kubernetes/kubernetes/releases/tag/v1.15.2
- https://github.com/kubernetes/kubernetes/releases/tag/v1.14.5
- https://github.com/kubernetes/kubernetes/releases/tag/v1.13.9

/close
